### PR TITLE
Release v0.3.0

### PR DESCRIPTION
**Pull request by pryorda**

### CODE REVIEW

**Comment by pryorda** [comment]
@dannyk81 This is some good stuff and you might be interested. I would also like to see if I could pull you in as a common reviewer?

**Comment by dannyk81** [comment]
Thanks @pryorda 😄 would be honored! sorry for being less responsive lately, holidays/vacations/family time...

I'm going to give it a try soon and report back if there's any issues but looks very exciting.

@jc2k really good stuff!!! 🙏 🚀

**Comment by pryorda** [comment]
@dannyk81 No worries. This is an open source project and not your paycheck :)

File: README.md [hunk 1]
```diff
@@ -150,61 +150,61 @@ You can use the following parameters in prometheus configuration file. The `para
...
+# HELP vmware_datastore_vms VMware Virtual Machines number using this datastore
 # TYPE vmware_datastore_vms gauge
 vmware_datastore_vms{ds_name="ESX1-LOCAL"} 0.0
 
-# HELP vmware_host_power_state VMWare Host Power state (On / Off)
+# HELP vmware_host_power_state VMware Host Power state (On / Off)
 # TYPE vmware_host_power_state gauge
 vmware_host_power_state{host_name="esx1.company.com"} 1.0
-# HELP vmware_host_cpu_usage VMWare Host CPU usage in Mhz
+# HELP vmware_host_cpu_usage VMware Host CPU usage in MHz
```
Comment (dannyk81) [inline README.md]: just nit-picking here, but correct abbreviation for megahertz is actually `MHz` (following: Hz, kHz, MHz, GHz notations in electronics)

Comment (dannyk81) [reply README.md]: ignore me... looks like I need to check my vision, but the z is indeed lower-case 🤦‍♂️

File: Dockerfile [hunk 2]
```diff
@@ -11,7 +11,7 @@ RUN pip install flake8 && \
 RUN flake8 vmware_exporter
 RUN dockerfilelint Dockerfile
 
-FROM python:2.7-alpine
+FROM python:3.6-alpine
 
 LABEL MAINTAINER="Daniel Pryor <daniel@pryorda.net>"
 LABEL NAME=vmware_exporter
```
Comment (dannyk81) [inline Dockerfile]: @pryorda let's bump the version here as well (Line 18)

**Comment by Jc2k** [comment]
Looks good to me!

Thanks for the quick turn around with a release! ❗️

**Comment by negast** [comment]
Hello 
I've tested this code with vmware/vcenter 6.7 python 3.5

I've encountered two issues 

First one being in the vmware_host_metadata method
It seems you require the datastore to be the smae name as the host_name which in my environment was not the case
```
def _vmware_host_metadata(self, inventory, host):
        """
        Get Host metadata from inventory
        """
        host_name = host.name
        host_dc_name = inventory[host_name]['dc']
        host_cluster_name = inventory[host_name]['cluster']

        return host_name, host_dc_name, host_cluster_name
```

I changed it to the following, where i take the first datastore of the host and search for it in the inventory I also had no cluster key so had to search for ds_cluster

```
def _vmware_host_metadata(self, inventory, host):
        """
        Get Host metadata from inventory
        """
        host_name = host.name
        ds_name = host.datastore[0].name
        host_dc_name = inventory[ds_name]['dc']
        if 'cluster' in inventory[ds_name]:
            host_cluster_name = inventory[ds_name]['cluster']
        else:
            host_cluster_name = inventory[ds_name]['ds_cluster']

        return host_name, host_dc_name, host_cluster_name
```

Second issue was an issue with appending tuples in the _vmware_get_vm_metrics module
I changed it to the following. I also created a new tuple for each specific disk else I always had the first disk path. I did the same for each snapshot.
```
def _vmware_get_vm_metrics(self, content, virtual_machine, metrics, inventory):
        """
        Get VM Guest Metrics
        """
        summary = virtual_machine.summary

        vm_metadata = self._vmware_vm_metadata(inventory, virtual_machine, summary)
        self._labels[virtual_machine._moId] = vm_metadata

        if self.collect_only['vms'] is True:
            vm_power_state = 1 if summary.runtime.powerState == 'poweredOn' else 0
            vm_num_cpu = summary.config.numCpu

            metrics['vmware_vm_power_state'].add_metric(vm_metadata, vm_power_state)
            metrics['vmware_vm_num_cpu'].add_metric(vm_metadata, vm_num_cpu)

            # Get metrics for poweredOn vms only
            if vm_power_state:
                if summary.runtime.bootTime:
                    metrics['vmware_vm_boot_timestamp_seconds'].add_metric(
                        vm_metadata,
                        self._to_epoch(summary.runtime.bootTime)
                    )

        if self.collect_only['vmguests'] is True:
            # gather disk metrics
            if len(virtual_machine.guest.disk) > 0:
                for disk in virtual_machine.guest.disk:
                    vm_metadata2 = (*vm_metadata, disk.diskPath)
                    metrics['vmware_vm_guest_disk_free'].add_metric(
                        vm_metadata2, disk.freeSpace)
                    metrics['vmware_vm_guest_disk_capacity'].add_metric(
                        vm_metadata2, disk.capacity)

        if self.collect_only['snapshots'] is True and virtual_machine.snapshot is not None:
            snapshots = self._vmware_full_snapshots_list(virtual_machine.snapshot.rootSnapshotList)

            metrics['vmware_vm_snapshots'].add_metric(
                vm_metadata,
                len(snapshots),
            )

            for snapshot in snapshots:
                vm_metadata3 = (*vm_metadata, snapshot['name'])
                metrics['vmware_vm_snapshot_timestamp_seconds'].add_metric(
                    vm_metadata3,
                    snapshot['timestamp_seconds'],
                )
```

After these changes the metrics showed up correctly.

**Comment by Jc2k** [comment]
Can I do anything to help here?

I agree with @dannyk81's suggested change for the first point.

It looks like we've already fixed the 2nd point (but a different way).

I'm not sure if there is a third issue or not - that sounds like its caused by the 2nd point.

**Comment by negast** [comment]
> Can I do anything to help here?
> 
> I agree with @dannyk81's suggested change for the first point.
> 
> It looks like we've already fixed the 2nd point (but a different way).
> 
> I'm not sure if there is a third issue or not - that sounds like its caused by the 2nd point.

My bad must've cloned an old version then. I won't be able to test again till monday but I'll try again then.

File: vmware_exporter/vmware_exporter.py [hunk 3]
```diff
@@ -281,50 +183,40 @@ def collect(self, vsphere_host, section='default'):
...
+            self.threader.submit(
+                self._vmware_get_datastores,
+                content, metrics, ds_inventory,
+            )
+
+        # Collect Hosts metrics
+        if collect_only['hosts'] is True:
+            self.threader.submit(
+                self._vmware_get_hosts,
+                content, metrics, ds_inventory,
```
[the remark this answers is not in the document]
Comment (pryorda) [reply vmware_exporter/vmware_exporter.py]: @Jc2k @dannyk81 I'm working on getting this cleaned up now.

**Comment by pryorda** [comment]
@negast I'm going to release v0.3.0. Everything looks good from my end. @dannyk81 Anything else you want looked at before I do?

**Comment by dannyk81** [comment]
lgtm @pryorda 👍

**Pull request merged**